src/pip/_internal/index.py

```diff
@@ -1333,9 +1336,17 @@ def _clean_link(url):
         path = urllib_request.pathname2url(
             urllib_request.url2pathname(result.path))
     else:
-        # In addition to the `/` character we protect `@` so that
-        # revision strings in VCS URLs are properly parsed.
-        path = urllib_parse.quote(urllib_parse.unquote(result.path), safe="/@")
+        # In addition to the `/` character we protect `@`
+        # (as well as theirs %-escapes) so that revision
+        # strings in VCS URLs are properly parsed.
+        path_bits = _safe_chars_re.split(result.path)
+        path_parts = [urllib_parse.quote(urllib_parse.unquote(path_bits[0]))]
+        for i in range(1, len(path_bits), 2):
```

Comment:
This may look cleaner if we operate on `iter(_safe_chars_re.split(result.path))`. So we would have `next(path_bits)` instead of `path_bits[0]` and in the loop we could use the `pairwise` [itertools recipe](https://docs.python.org/2/library/itertools.html#recipes) to get element pairs.

Reply:
I'm afraid i'm not following how to do it. Could you write a sample snippet?

Reply:
Sure, I was thinking something like this:

```
path_bits = iter(_safe_chars_re.split(result.path))
path_parts = [urllib_parse.quote(urllib_parse.unquote(next(path_bits)))]
for preserved, to_quote in pairwise(path_bits):
    path_parts.append(preserved)
    path_parts.append(urllib_parse.quote(urllib_parse.unquote(to_quote)))
```

Reply:
I see. Due to doc `pairwise(s)` produces `(s0,s1), (s1,s2), (s2, s3), ...`, but forloop is expecting `(s0,s1), (s2,s3), (s4, s5),`.

Reply:
Oops, you're right - in this case we can use `izip(path_bits, path_bits)` or wrap it in our own non-overlapping pairwise like

```python
def pairwise(iterable):
    iterable = iter(iterable)
    return izip(iterable, iterable)

Reply:
Also, it would be better to use a pattern that doesn't require repeating `urllib_parse.quote(urllib_parse.unquote())` twice, once before the loop and again during the loop. A couple options for doing that include adding an empty string to make the list have an even number of elements and perhaps using `itertools.zip_longest()`.

Reply:
Nice idea, something like `pairwise(chain(path_bits, ['']))` would work.